Add additional asserts to help investigate test_win_pkg failure

Ch3LL opened the pull request:
### What does this PR do?
Currently the `test_win_pkg` module integration tests is failing sometimes on windows. I cannot replicate this failure so I'm adding additional asserts and running the `refresh_db` one more time if it does initially fail.

Ch3LL committed:
Add additional asserts to help investigate test_win_pkg failure

cachedout commented on tests/integration/modules/test_win_pkg.py:
```diff
@@ -33,8 +34,24 @@ def test_adding_removing_pkg_sls(self):
         '''
-        def _check_pkg(pkgs, exists=True):
-            self.run_function('pkg.refresh_db')
+        def _check_pkg(pkgs, exists=True, check_refresh=None):
```
Should `check_refresh` be a boolean by default?

Ch3LL replied:
I could change it to boolean but we are passing in `check_refresh=2`, when using this so i kept it at None. Is it better practice to keep it a boolean even though we are not passing True/False into the kwarg.

damon-atkins commented on tests/integration/modules/test_win_pkg.py:
```diff
@@ -33,8 +34,24 @@ def test_adding_removing_pkg_sls(self):
+                        self.assertEqual(check_refresh, refresh['total'],
+                             msg='total returned {0}. Expected return {1}'.format(check_refresh, refresh['total']))
+                        self.assertEqual(check_refresh, refresh['success'],
+                             msg='success returned {0}. Expected return {1}'.format(check_refresh, refresh['success']))
```
msg='success returned {0}. Expected return {1}'.format(refresh['success']), check_refresh)

damon-atkins commented on tests/integration/modules/test_win_pkg.py:
```diff
@@ -33,8 +34,24 @@ def test_adding_removing_pkg_sls(self):
+                            raise AssertionError(err)
+                        count = count -1
+                        refresh = self.run_function('pkg.refresh_db')
             repo_data = self.run_function('pkg.get_repo_data', timeout=300)
```
timeout=30   it just opens an existing file and reads it into ram.  30 seconds is heaps

Ch3LL replied:
looks like that change was made intentionally to also try to stabilize this flaky test so I dont want to remove it quit yet.

damon-atkins replied:
You could also call pkg.refresh_db with verbose=True and failhard=False
The data structure return will be this even if it fails (with data not like the example below). 
```
{
    "local": {
        "total": 0,
        "success": 0,
        "success_list": {},
        "failed_list": {},
        "failed": 0
    }
}
```

Ch3LL committed:
Fix the assertion message order

dwoz committed:
Debug win_pkg

dwoz commented:
@Ch3LL I'vd made some additions to this: https://github.com/Ch3LL/salt/pull/4

Ch3LL committed:
Merge pull request #4 from dwoz/win_pkg_flaky

Fix up win_pkg test

dwoz committed:
Log results of cache_dir for now

Ch3LL committed:
Merge pull request #5 from dwoz/win_pkg_flaky

Log results of cache_dir for now

dwoz committed:
Try running fsync to avoid fs caching race condition

Ch3LL committed:
Merge pull request #6 from dwoz/win_pkg_flaky

Try running fsync to avoid fs caching race condition

damon-atkins commented on salt/modules/win_pkg.py:
```diff
@@ -693,12 +693,13 @@ def refresh_db(**kwargs):
         include_pat='*.sls',
         exclude_pat=r'E@\/\..*?\/'  # Exclude all hidden directories (.git)
     )
+    log.debug("refresh_db - Return from cache_dir %s", repr(ret))
```
Not sure why this is needed, it already spills out heaps of logs in debug. i.e. every file which is pulled down and every file skipped.

damon-atkins commented:
Just looking at the windows test and there are hundreds of these lines `00:39:41 unix2dos: converting file`
Which is not how it works when its running for real, being installed normally.

dwoz committed:
Fix win_pkg test

Bugfix in file_list cache. The file_list cache age could be a negative
number meaning the file_list cache can get used in tests even though the
cache timeout is set to 0 for test suite runs.

Ch3LL committed:
Merge pull request #7 from dwoz/win_pkg_flaky

Fix win_pkg test

dwoz commented:
@damon-atkins We added the unix2dos conversion a month or two ago. This was done because, by default, git will perform newline translations and git clones on Windows will have CRLF line endings. By running unix2dos the tests run with the same line endings as if someone runs the tests from a git clone on windows.

Due to the fact that git can be configured to perform newline translations (or not), using unix2dos is not ideal. Since you could perform a git clone without newline translations and then see some failures in the test suite because of that.

In addition, as you have pointed out, using unix2dos does not cover the case where you have a linux master with windows minion. It means the test suite primarily testing a Windows master with a Windows minion.

I think long term we want to enumerate all possible line ending cases and test all of them. We're just not there yet. A couple other things to think about is the fact that YAML performs line ending translation so that values parsed by YAML always have LF line endings even if the source file has CRLF line endings. JSON on the other hand *should* always have LF line endings source files.

Hopefully this explains why we are using unix2dos for now and the direction we are heading.

damon-atkins commented:
Thanks.   And assume pip installs would have the original line endings as well from the source.
